sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/NonAsyncTaskShouldNotReturnNullBase.cs

```diff
@@ -0,0 +1,63 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class NonAsyncTaskShouldNotReturnNullBase : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4586";
+        protected const string MessageFormat = "Do not return null from this method, instead return 'Task.FromResult<T>(null)', " +
+            "'Task.CompletedTask' or 'Task.Delay(0)'.";
+
+        private static readonly ISet<KnownType> TaskTypes =
+            new HashSet<KnownType>
+            {
+                KnownType.System_Threading_Tasks_Task,
+                KnownType.System_Threading_Tasks_Task_T
+            };
+
+        protected static bool IsInvalidEnclosingSymbolContext(SyntaxNode enclosingMember, SemanticModel model)
+        {
+            var enclosingMemberSymbol = model.GetDeclaredSymbol(enclosingMember);
```

Comment:
Wouldn't it be better to cast to `IMethodSymbol` only one, here, instead of doing it in the next two following functions?